.github: Add workflow to stale pull requests

seemethere opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51237 .github: Add workflow to stale pull requests**

Stales pull requests at 150 days and then will close them at 180 days

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

Differential Revision: [D26112086](https://our.internmc.facebook.com/intern/diff/D26112086)

seemethere committed:
.github: Add workflow to stale pull requests

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on ".github: Add workflow to stale pull requests"


Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

samestep reviewed (changes requested):
As written, this seems like it will mark both PRs _and_ issues as stale after six months, rather than only PRs; is that correct? If so, it should be changed to only act on PRs, not issues.

yns88 commented:
Is it better to do this as our own GitHub Action, or to use the Stale app that's already installed in this org?

yns88 commented on .github/workflows/stale_pull_requests.yml:
```diff
@@ -0,0 +1,20 @@
+            Looks like this PR hasn't been updated in a while! Going to go ahead and mark this as `stale`.
+            Feel free to update / remove the `stale` label if you feel this is a mistake
+            `stale` pull requests will automatically be closed 30 days after being marked `stale`
+          exempt-pr-labels: "no-stale"
```
If we add "open source" and "high priority" exemptions, then I think we can get more aggressive and close after as little as 3 months.

seemethere replied:
I've added a separate job specifically for open source pull requests only

seemethere commented:
> As written, this seems like it will mark both PRs _and_ issues as stale after six months, rather than only PRs; is that correct? If so, it should be changed to only act on PRs, not issues.

The [usage docs](https://github.com/actions/stale) from the repository indicate that if no message for issues is designated then issues will not be marked stale.

samestep reviewed (approved):
> The [usage docs](https://github.com/actions/stale) from the repository indicate that if no message for issues is designated then issues will not be marked stale.

Ah thanks, I missed that!

seemethere commented:
> Is it better to do this as our own GitHub Action, or to use the Stale app that's already installed in this org?

I think using an action for this is better in the long run since it makes it transparent how / when it's called.

seemethere committed:
Update on ".github: Add workflow to stale pull requests"


Stales pull requests at 150 days and then will close them at 180 days

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

seemethere committed:
Update on ".github: Add workflow to stale pull requests"


Stales pull requests at 150 days and then will close them at 180 days

Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

[ghstack-poisoned]

walterddr commented on .github/workflows/stale_pull_requests.yml:
```diff
@@ -0,0 +1,34 @@
+name: 'Close stale pull requests'
+on:
+  schedule:
+    - cron: '30 1 * * *'
```
nit: do we need to run it so frequently :-P

walterddr replied:
actually nvm. doesn't hurt since it is so lightweight

yns88 replied:
Isn't this just daily? That seems like the minimum viable frequency, since this will probably start closing multiple PRs per day after it gets through the backlog.

Actually at the default operations-per-run value of 30, we'll need to run this workflow dozens of times in a row to get through the backlog.

walterddr replied:
yeah I was thinking since we are doing a 30-60, might as well just do weekly or sth during the weekend.

yns88 commented:
> I think using an action for this is better in the long run since it makes it transparent how / when it's called.

Seems legit. Plus doing it this way lets us to different retentions for different sets of PRs, which doesn't seem possible with the Stale app.
